### DISCUSSION TITLE
Add a preferred/favorite professor for each course

### DISCUSSION THREAD

**Discussion by dlutton** [MattStrauss/sweng-scheduler#7] [Ideas]
For each course, it would be nice if we could see a preferred or favorite professor based on some type of voting system.

**Comment by MattStrauss** [comment]
We would have to add a Professor model which would have a many-to-many relationship with the Course model. Then, if we are going to set up a voting system, it might be smart to make it polymorphic as I imagine we will want to vote on other things as well. Maybe make a `Votable` trait or something like that...

**Reply by MwesterlundDev** [reply]
Another thought could be to do something like "up-vote" or "down-vote."  And have something like an aggregate score for professors, courses, etc.

**Comment by dlutton** [comment]
For the backend we can probably keep it simple with a professor table with 4 columns (id (key), professor name, course (maybe foreign key (haven't looked at the db yet), and votes).  We can allow students to add their favorite professor based on a pre-populated list which will add 1 to that votes column).  On the UI we can simply display the professor with the highest vote count as a favorite for that course.

**Reply by dlutton** [reply]
obviously if a student changes their favorite for that course, it'll subtract 1 from the one they had previously and add 1 to the new one.  Maybe 2 routes or methods (unsure of the current architecture), 1 to add your favorite which takes in the professor's id and another to modify which takes in the current one and new one (new one is optional here in the event the student doesn't want to select any professor as a new favorite - I'm not familiar with php so either an optional param or overloaded method.)

**Reply by MwesterlundDev** [reply]
I think on the UI we could do something like list the professors for each class, with their vote scores.  Often times I've found the availability of the course dictates when the course is taken, and you sort of just have to swallow the pill of a difficult prof.  But it would be nice going into the semester at least knowing that.

**Reply by MattStrauss** [reply]
Just realized another issue with this today when implementing the Professors model and then the Course_Professor Lookup table... many classes only have one Prof. so the voting may not make sense in these scenarios. See [below comment about an alt strategy for voting

**Comment by MwesterlundDev** [comment]
In general, one thing I think we should stay away from is any free formed text. I don't think we want to get into the business of monitoring comments on professors or classes, etc.  I think something like up-vote and down-vote will give enough info.  Another option could be a 5 star system like amazon or something.

**Reply by MattStrauss** [reply]
Totally agree on this. I was thinking just a simple Select your fav Prof/Course from the entire list of courses

**Comment by MattStrauss** [comment]
I began implementation of the voting system with polymorphism and a "votable" trait. But as I was thinking about it more, it might make more sense to create a Polls section, whereas we can post different poll questions and allow users to vote on them. While this is somewhat outside the overall scope of the project it would be a fun way to keep users involved. 

Some thoughts for polls would be like our original ones: favorite course, favorite professor, etc. But maybe we can allot the user 50 points (or units or w/e) and they can use all 50 on one course/prof or spread the points out over a few different selections. It's just a thought. We could also use a star rating or something like that as well. 

Overall the Polls portion of the app wouldn't be very integrated with the main Course Viewing and Scheduling functionality, except to show the occasional results in a given view when applicable.
